### PR TITLE
Add support for dump on CX6DX NIC

### PR DESCRIPTION
**Pull request by muhamadSammar**
Adding parsers for CX6DX STE's and tag formats
Signed-off-by: Muhammad Sammar <muhammads@mellanox.com>

### CODE REVIEW

File: src/dr_rule.py [hunk 1]
```diff
@@ -104,14 +104,22 @@ def __init__(self, data):
...
+        parsed_ste = mlx5_hw_ste_parser(nic_version, self.data['ste_data'], raw, verbose)
         if "tag" not in parsed_ste.keys():
             return ""
 
         if verbose == 0:
             return "%s" % dict_join_str(parsed_ste["tag"])
         else:
-            if int(self.data['dr_dump_rec_type']) == dr_dump_rec_type.DR_DUMP_REC_TYPE_RULE_RX_ENTRY.value[0]:
+            if int(self.data['dr_dump_rec_type']) in [dr_dump_rec_type.DR_DUMP_REC_TYPE_RULE_RX_ENTRY_V0.value[0],
+                                                      dr_dump_rec_type.DR_DUMP_REC_TYPE_RULE_RX_ENTRY_V0.value[0]]:
```
Comment (dariavel) [inline src/dr_rule.py:122]: did you mean V1?

File: src/dr_utilities.py [hunk 2]
```diff
@@ -36,6 +36,58 @@
 g_indent = 0
 TAB = "   "
 
+
+class mlx5_ifc_steering_format_version(Enum):
```
Comment (dariavel) [inline src/dr_utilities.py:40]: remove Enum dependancy

File: src/parsers/dr_ste_parser.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from src.dr_utilities import mlx5_ifc_steering_format_version
+from src.parsers.dr_ste_v0_parser import mlx5_hw_ste_v0_parser
+from src.parsers.dr_ste_v1_parser import mlx5_hw_ste_v1_parser
+
+def mlx5_hw_ste_parser(nic_version, hex_str, raw, verbose):
+    arr = {
```
Comment (dariavel) [inline src/parsers/dr_ste_parser.py:6]: replace with:
`x = str(bin(0x7))[2:]`
`y =( 4-len(x))*'0' + x`

File: src/parsers/dr_ste_parser.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        "0": "0000", "1": "0001", "2": "0010", "3": "0011",
+        "4": "0100", "5": "0101", "6": "0110", "7": "0111",
+        "8": "1000", "9": "1001", "a": "1010", "b": "1011",
+        "c": "1100", "d": "1101", "e": "1110", "f": "1111"
+    }
+
+    bin_str = ""
+
+    for i in range(0, len(hex_str)):
```
Comment (dariavel) [inline src/parsers/dr_ste_parser.py:15]: Try using this instead
 `bin(int(ste, 16))`

File: src/parsers/dr_ste_parser.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        bin_str += arr[hex_str[i]]
+
+    if nic_version == mlx5_ifc_steering_format_version.MLX5_HW_CONNECTX_5.value[0]:
+        return mlx5_hw_ste_v0_parser(bin_str, raw)
+    elif nic_version == mlx5_ifc_steering_format_version.MLX5_HW_CONNECTX_6DX.value[0]:
+        return mlx5_hw_ste_v1_parser(bin_str, raw, verbose)
+    else:
+        print("Unsupported Nic version, currently supporting CX5 and CX6DX")
```
Comment (dariavel) [inline src/parsers/dr_ste_parser.py:23]: at least "NIC" better "device"

File: src/parsers/dr_ste_v0_parser.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+#        provided with the distribution.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
```
Comment (dariavel) [inline src/parsers/dr_ste_v0_parser.py:29]: check the copiright; add everywhere if needed, or delete everywhere if not.

File: src/parsers/dr_ste_v0_parser.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+from src.parsers.dr_ste_v0_tag_parser import mlx5_ste_v0_tag_parser
+
+def conv_ip_version(version):
```
Comment (dariavel) [inline src/parsers/dr_ste_v0_parser.py:33]: take those 3 functions from utils

File: src/parsers/dr_ste_v0_parser.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    ret["match_polarity"] = _val(bin_str[252: 253])
+    ret["mask_mode"] = _val(bin_str[253: 254])
+    ret["miss_rank"] = _val(bin_str[254: 256])
+    ret["tag"] = mlx5_ste_v0_tag_parser(ret["entry_sub_type"], bin_str[256: 384], raw)
+    return ret
+
+def mlx5_hw_ste_v0_parser(bin_str, raw) :
+	entry_type = int(bin_str[0: 4], 2)
+	switch = { 1 : mlx5_ifc_ste_v0_sx_transmit_bits_parser,
+			   2 : mlx5_ifc_ste_v0_rx_steering_mult_bits_parser,
```
Comment (dariavel) [inline src/parsers/dr_ste_v0_parser.py:164]: fix indentation

File: src/parsers/dr_ste_v0_parser.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+			   6 : mlx5_ifc_ste_v0_modify_packet_bits_parser
+	}
+
+	return switch[entry_type](bin_str, raw, )
```
Comment (dariavel) [inline src/parsers/dr_ste_v0_parser.py:168]: remove "," coma after raw

File: src/parsers/dr_ste_v0_parser.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    ret["miss_address_31_6"] = _val(bin_str[224: 250])
+    ret["learning_point"] = _val(bin_str[250: 251])
+    ret["fail_on_error"] = _val(bin_str[251: 252])
+    ret["match_polarity"] = _val(bin_str[252: 253])
+    ret["mask_mode"] = _val(bin_str[253: 254])
+    ret["miss_rank"] = _val(bin_str[254: 256])
+    ret["tag"] = mlx5_ste_v0_tag_parser(ret["entry_sub_type"], bin_str[256: 384], raw)
+    return ret
+
+def mlx5_hw_ste_v0_parser(bin_str, raw) :
```
Comment (dariavel) [inline src/parsers/dr_ste_v0_parser.py:161]: remove space before :

File: src/parsers/dr_ste_v0_tag_parser.py [hunk 11]
```diff
@@ -523,138 +505,4 @@ def mlx5_tag_parser(lookup_type, tag, raw):
...
-        bin_str += arr[hex_str[i]]
-
-    entry_type = int(bin_str[0: 4], 2)
-    switch = {1: mlx5_ifc_ste_sx_transmit_bits_parser,
-              2: mlx5_ifc_ste_rx_steering_mult_bits_parser,
-              6: mlx5_ifc_ste_modify_packet_bits_parser
-              }
-
-    return switch[entry_type](bin_str, raw)
+    return parsed_tag
```
Comment (dariavel) [inline src/parsers/dr_ste_v0_tag_parser.py:508]: add new line at the end of file

File: src/parsers/dr_ste_v1_parser.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    return ret
+
+
+def mlx5_hw_ste_v1_parser(bin_str, raw, verbose):
+    entry_type = int(bin_str[0: 8], 2)
+    switch = {
+        dr_ste_entry_format.DR_STE_TYPE_BWC_BYTE.value[0] : mlx5_ifc_ste_v1_match_bwc_byte_bits_parser,
+        dr_ste_entry_format.DR_STE_TYPE_BWC_DW.value[0] : mlx5_ifc_ste_v1_match_bwc_byte_bits_parser,
+        dr_ste_entry_format.DR_STE_TYPE_MATCH.value[0] : mlx5_ifc_ste_v1_match_bwc_dw_bits_parser,
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_parser.py:84]: Fix naming

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+	ret["reserved_at_21"] = _val(bin_str[33: 34])
+	ret["source_is_requestor"] = _val(bin_str[34: 35])
+	ret["reserved_at_23"] = _val(bin_str[35: 40])
+	ret["source_qp"] = _val(bin_str[40: 64])
+	ret["reserved_at_40"] = _val(bin_str[64: 96])
+	ret["reserved_at_60"] = _val(bin_str[96: 128])
+	return ret
+
+
+def mlx5_ifc_ste_icmp_v1_bits_tag_parser(bin_str):
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:370]: Remove unused

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 14]
```diff
@@ -0,0 +1,424 @@
+from src.dr_utilities import dr_ste_v1_lu_type
+from src.dr_utilities import _val
+from src.parsers.dr_ste_v0_tag_parser import *
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:3]: list the 4 that you are using instead of "*"

File: src/parsers/dr_ste_parser.py (hunk 3 above)
Comment (dariavel) [reply src/parsers/dr_ste_parser.py]: Wating for the next commit with this

File: src/parsers/dr_ste_parser.py (hunk 4 above)
Comment (dariavel) [reply src/parsers/dr_ste_parser.py]: the same

File: src/parsers/dr_ste_v1_tag_parser.py (hunk 14 above)
Comment (dariavel) [reply src/parsers/dr_ste_v1_tag_parser.py]: from src.parsers.dr_ste_v0_tag_parser import mlx5_ifc_ste_v0_general_purpose_bits_tag_parser,  mlx5_ifc_ste_v0_eth_l3_ipv6_dst_bits_tag_parser_p, mlx5_ifc_ste_v0_mpls_bits_tag_parser, mlx5_ifc_ste_v0_register_0_bits_tag_parser, mlx5_ifc_ste_v0_register_1_bits_tag_parser, mlx5_ifc_ste_v0_eth_l3_ipv6_src_bits_tag_parser_p

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 15]
```diff
@@ -0,0 +1,450 @@
...
+from src.dr_utilities import add_inner_to_key
+import src.dr_prettify as dr_prettify
+from src.parsers.dr_ste_v0_tag_parser import mlx5_ifc_ste_v0_general_purpose_bits_tag_parser
+from src.parsers.dr_ste_v0_tag_parser import mlx5_ifc_ste_v0_eth_l3_ipv6_dst_bits_tag_parser_p
+from src.parsers.dr_ste_v0_tag_parser import mlx5_ifc_ste_v0_mpls_bits_tag_parser
+from src.parsers.dr_ste_v0_tag_parser import mlx5_ifc_ste_v0_register_0_bits_tag_parser
+from src.parsers.dr_ste_v0_tag_parser import mlx5_ifc_ste_v0_register_1_bits_tag_parser
+from src.parsers.dr_ste_v0_tag_parser import mlx5_ifc_ste_v0_eth_l3_ipv6_src_bits_tag_parser_p
+
+def mlx5_ifc_ste_eth_l2_src_v1_bits_tag_parser(bin_str) :
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:42]: not used

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 16]
```diff
@@ -0,0 +1,450 @@
...
+	ret["l3_ok"] = _val(bin_str[108: 109])
+	ret["l4_ok"] = _val(bin_str[109: 110])
+	ret["second_vlan_qualifier"] = _val(bin_str[110: 112])
+	ret["second_priority"] = _val(bin_str[112: 115])
+	ret["second_cfi"] = _val(bin_str[115: 116])
+	ret["second_vlan_id"] = _val(bin_str[116: 128])
+	return ret
+
+
+def mlx5_ifc_ste_eth_l2_dst_v1_bits_tag_parser(bin_str) :
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:105]: not used

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 17]
```diff
@@ -0,0 +1,450 @@
...
+	ret["l3_type"] = _val(bin_str[90: 92])
+	ret["reserved_at_5c"] = _val(bin_str[92: 94])
+	ret["first_vlan_qualifier"] = _val(bin_str[94: 96])
+	ret["first_priority"] = _val(bin_str[96: 99])
+	ret["first_cfi"] = _val(bin_str[99: 100])
+	ret["first_vlan_id"] = _val(bin_str[100: 112])
+	return ret
+
+
+def mlx5_ifc_ste_eth_l3_ipv4_5_tuple_v1_bits_tag_parser(bin_str) :
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:204]: not used

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 18]
```diff
@@ -0,0 +1,450 @@
...
+	ret["tcp_rst"] = _val(bin_str[109: 110])
+	ret["tcp_syn"] = _val(bin_str[110: 111])
+	ret["tcp_fin"] = _val(bin_str[111: 112])
+	ret["dscp"] = _val(bin_str[112: 118])
+	ret["ecn"] = _val(bin_str[118: 120])
+	ret["protocol"] = _val(bin_str[120: 128])
+	return ret
+
+
+def mlx5_ifc_ste_eth_l2_tnl_v1_bits_tag_parser(bin_str) :
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:254]: not used

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 19]
```diff
@@ -0,0 +1,450 @@
...
+	ret["window_size"] = _val(bin_str[0: 16])
+	ret["urgent_pointer"] = _val(bin_str[16: 32])
+	ret["ack_num"] = _val(bin_str[32: 64])
+	ret["seq_num"] = _val(bin_str[64: 96])
+	ret["length"] = _val(bin_str[96: 112])
+	ret["checksum"] = _val(bin_str[112: 128])
+	return ret
+
+
+def mlx5_ifc_ste_mpls_v1_bits_tag_parser(bin_str) :
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:350]: not used

File: src/parsers/dr_ste_v1_tag_parser.py (hunk 13 above)
Comment (dariavel) [reply src/parsers/dr_ste_v1_tag_parser.py]: still found more that are not used..

File: src/parsers/dr_ste_v1_tag_parser.py (hunk 17 above)
Comment (muhamadSammar) [reply src/parsers/dr_ste_v1_tag_parser.py:204]: the ..._p (pretty) is used, usually we keep both structs

Comment (dariavel) [reply src/parsers/dr_ste_v1_tag_parser.py:204]: what's the difference?

File: src/parsers/dr_ste_v1_parser.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+    ret["action1"] = hex(int(bin_str[224: 256], 2))
+
+    tag = bin_str[256 : 384]
+    lookup_type = int(bin_str[0 : 8] + bin_str[48 : 56], 2)
+    ret["tag"] = mlx5_ste_v1_tag_parser(lookup_type, tag, raw)
+
+    return ret
+
+
+def mlx5_ifc_ste_v1_match_bits_parser(bin_str, raw) :
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_parser.py:74]: raw parameter is not used

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+from src.dr_utilities import _val
+from src.dr_utilities import add_inner_to_key
+import src.dr_prettify as dr_prettify
+from src.parsers.dr_ste_v0_tag_parser import mlx5_ifc_ste_v0_general_purpose_bits_tag_parser,\
+	mlx5_ifc_ste_v0_eth_l3_ipv6_dst_bits_tag_parser_p, mlx5_ifc_ste_v0_mpls_bits_tag_parser,\
+	mlx5_ifc_ste_v0_register_0_bits_tag_parser, mlx5_ifc_ste_v0_register_1_bits_tag_parser,\
+	mlx5_ifc_ste_v0_eth_l3_ipv6_src_bits_tag_parser_p
+
+
+def mlx5_ifc_ste_eth_l2_src_v1_bits_tag_parser_p(bin_str) :
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:41]: remove spaces before ":"
this function definition and several more below

File: src/dr_utilities.py [hunk 22]
```diff
@@ -241,11 +241,8 @@ def add_inner_to_key(in_dict):
...
-    arr = {"0": "0000", "1": "0001", "2": "0010", "3": "0011",
-           "4": "0100", "5": "0101", "6": "0110", "7": "0111",
-           "8": "1000", "9": "1001", "a": "1010", "b": "1011",
-           "c": "1100", "d": "1101", "e": "1110", "f": "1111"}
-    bin_str = ""
-    for i in range(0, len(hex_str)):
-        bin_str += arr[hex_str[i]]
+    #To save the first zeroes from being compressed by 'bin'
+    hex_str = 'f' + hex_str
+    #convert to binray and remove "0b1111"
```
Comment (dariavel) [inline src/dr_utilities.py:246]: binray -> binary

File: src/parsers/dr_ste_v0_parser.py [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from src.parsers.dr_ste_v0_tag_parser import mlx5_ste_v0_tag_parser
+from src.dr_utilities import conv_ip_version, _val, add_inner_to_key
```
[the remark this answers is not in the document]
Comment (muhamadSammar) [reply src/parsers/dr_ste_v0_parser.py:32]: done

File: src/parsers/dr_ste_v1_tag_parser.py (hunk 21 above)
Comment (muhamadSammar) [reply src/parsers/dr_ste_v1_tag_parser.py:41]: done

File: src/parsers/dr_ste_v1_parser.py (hunk 20 above)
Comment (muhamadSammar) [reply src/parsers/dr_ste_v1_parser.py:74]: i need it for line 117

Comment (dariavel) [reply src/parsers/dr_ste_v1_parser.py:74]: but not in this function.. 117 is a different function

**Pull request merged**